proxy/coco/src/request/waiting_room.rs

```diff
@@ -0,0 +1,748 @@
+    ///   * If the underlying `Request` was not in the expected state.
+    ///   * If the underlying `Request` timed out.
+    pub fn queried(&mut self, urn: &RadUrn, timestamp: T) -> Result<(), Error>
+    where
+        T: Clone,
+    {
+        let max_queries = self.config.max_queries;
+        let max_clones = self.config.max_clones;
+        self.transition(
+            |request| match request {
+                SomeRequest::Created(request) => Some(Either::Right(request.request(timestamp))),
+                SomeRequest::Requested(request) => {
+                    Some(request.queried(max_queries, max_clones, timestamp))
+                },
+                _ => None,
+            },
+            |previous| match previous {
+                Either::Left(timeout) => Err(timeout.into()),
+                Either::Right(request) => Ok(request),
+            },
+            urn,
+        )
+    }
+
+    /// Tell the `WaitingRoom` that a `peer` was found for the given `urn`.
+    ///
+    /// If the underlying `Request` was in the `IsRequested` state then it will transition to the
+    /// `Found` state.
+    ///
+    /// If the underlying `Request` was in the `Found` or `Cloning` state then it add this `peer`
+    /// to the set of found peers.
+    ///
+    /// # Errors
+    ///
+    ///   * If the `urn` was not in the `WaitingRoom`.
+    ///   * If the underlying `Request` was not in the expected state.
+    pub fn found(&mut self, urn: &RadUrn, peer: PeerId, timestamp: T) -> Result<(), Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Requested(request) => Some(request.first_peer(peer, timestamp).into()),
+                SomeRequest::Found(request) => {
+                    let some_request: SomeRequest<T> = request.found(peer, timestamp).into();
+                    Some(some_request)
+                },
+                SomeRequest::Cloning(request) => {
+                    let some_request: SomeRequest<T> = request.found(peer, timestamp).into();
+                    Some(some_request)
+                },
+                _ => None,
+            },
+            Ok,
+            urn,
+        )
+    }
+
+    /// Tell the `WaitingRoom` that we failed at finding a peer to clone from for the given `urn`.
+    ///
+    /// This should mean that the set of peers in the `Found` state have been exhausted and are all
+    /// `Status::Failed`. It left up to the caller to determine this by using the
+    /// `HasPeers::all_failed` method.
+    ///
+    /// If the underlying `Request` was in the `Found` state then it will transition to the
+    /// `IsRequested` state.
+    ///
+    /// # Errors
+    ///
+    ///   * If the `urn` was not in the `WaitingRoom`.
+    ///   * If the underlying `Request` was not in the expected state.
+    pub fn found_failed(&mut self, urn: &RadUrn, timestamp: T) -> Result<(), Error>
```

Comment:
This should be collapsable into `cloning_failed` if that would take the `PeerId`. Otherwise the question would rise how does the caller keep track of that all found peers have been exhausted.

Reply:
Scratch that, totally miss-understood. My real question is, when is that method used?

Reply:
After all the effort I put into docs :(

```rust
    /// This should mean that the set of peers in the `Found` state have been exhausted and are all
    /// `Status::Failed`. It left up to the caller to determine this by using the
    /// `HasPeers::all_failed` method.
```

In other words, if we're in the `Found` state and have exhausted all our options then we need a way to fail back to `IsRequested` so that the caller can say `queried` again.

In terms of our actual use case, the `providers` stream has been emptied and none of the clone attempts worked so we need to call `providers` again to make another attempt.

Reply:
We discussed on a call and came to the conclusion that if the `Cloning` `failed` _and_ all the peers have failed then we can go back to `IsRequested`.

Reply:
As well as that, `IsRequested` should hold a `HashMap` of peers so that it doesn't forget what peers it has tried.